### ISSUE TITLE
Not able to update UI from events because of background thread

### ISSUE THREAD

**Issue by davecoffin** [NativeScript/nativescript-background-http#39]
I am using background-http to upload images to a server. I am using task.on to log events, and call callbacks based on the status. For example:

```
// from photos-view-model.js
task.on("progress", doEvent);
task.on("error", doEvent);
task.on("complete", doEvent);
task.on("responded", doEvent);
function doEvent(e) {
    var rtn = {
        "filename": filename,
        "status": e.eventName
    }
    if (e.eventName == 'complete') {
        console.log(e.eventName);
        if (cleanUp) cleanUp(rtn);
    } else if (e.eventName == 'progress') {
        if (updateProgressFn) updateProgressFn({uploaded: e.object.upload, total: e.object.totalUpload}, filename);
    } else if (e.eventName == 'responded') {
        rtn.data = e.data;
        if (callBack) callBack(rtn);
    }
}
```

I pass those callbacks from my main codebehind file:

```
// from photos.js
pageData.set('photoUploading', true);

var completedCallback = function(data) {
    console.log('the photo is done uploading.');
}

var cleanUp = function(data) {
    console.log('******************************');
    console.log('Setting photoUploading to false!');
    console.log('******************************');
    pageData.set('photoUploading', false);
}

var updateProgressFn = function(progress, filename) {
    var percent = (progress.uploaded*100)/progress.total;
    pageData.set('percentDone', percent.toFixed(0) + '%');
    console.log(percent.toFixed(0) + '%');
}
```

The logging works great. I can watch the percentage go up nicely, but it is not reflected in the UI. Here is my XML for testing:

```
<StackLayout id="uploading_holder">
    <Label style.color="white" text="{{'photoUploading: ' + photoUploading}}" />
    <Label style.color="white" text="{{'percentDone: ' + percentDone}}" />
    <Image class="uploadingImage" id="uploadingImage" />
    <ActivityIndicator busy="true" marginTop="20" color="white" marginBottom="10"/>
    <Label text="{{'Uploading ' + percentDone}}" id="uploading_text" class="mont" style.color="white" horizontalAlignment="center" marginBottom="20" />
</StackLayout>
```

The labels in the UI do not reflect the actual values being updated in my observable. I am doing it this way (just updating observables based on bhttp events) because when I try to update the UI directly like 

`page.getViewById('uploading_text').text = percentDone`

I got this error:

`This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.`

I assume the same issue is plaguing this approach, I am just not getting the error anymore. The behavior is the same...the UI is not consistently updated. Sometimes the percentage number will jump up, and sometimes eventually the UI will get the update, usually after about 15-45 seconds after I update the observable.

I've created a video for you to see the console output and contrast it against the view in the app. Notice at :13 in the video, "photoUploading" is set to false. then at :27, the UI is updated. 

https://cl.ly/i5F1

**Comment by Daxito** [comment]
+1 
I had the same issue! I added a workaround, assigned the Percent value to a global variable and then I had a timer that would check this global variable and then I would update the UI.... not perfect but works

**Comment by davecoffin** [comment]
@Daxito Thanks for that. I just implemented that workaround and it does the job :) You've eased my mind for now. The issue bugs me though, creates a lot more work to implement this plugin throughout my app.

**Comment by PanayotCankov** [comment]
This will be resolved in 2.4.0 we are about to publish
